### PR TITLE
[FLINK-10353][kafka] Support change of transactional semantics in Kaf…

### PR DESCRIPTION
**Pull request by StefanRRichter**
…ka Producer with existing state

## What is the purpose of the change

This PR changes `FlinkKafkaProducer` and  `FlinkKafkaProducer011` to support a change of transactional semantics when restoring from existing state, e.g. a savepoint.

## Brief change log

- Introduced `KafkaTransactionState#isTransactional` to distinguish which transactional handling should be applied instead of relying on what is currently configured.

- Call `initializeUserContext` in `TwoPhaseCommitSinkFunction` for all cases that did not recover a user context.

- Consider removing tranactional id from the properties when creating a new producer to deactivate transactional semantics if no longer required.

## Verifying this change

Added `FlinkKafkaProducer(11)ITCase#testMigrateFromAtLeastOnceToExactlyOnce` and `#testMigrateFromAtExactlyOnceToAtLeastOnce`

I wonder if we should also add tests from/to the NONE semantics. The tests are pretty heavy weight for what they are doing.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java [hunk 1]
```diff
@@ -957,6 +928,11 @@ private void recycleTransactionalProducer(FlinkKafkaProducer<byte[], byte[]> pro
 		return initProducer(registerMetrics);
 	}
 
+	private FlinkKafkaProducer<byte[], byte[]> initNonTransactionalProducer(boolean registerMetrics) {
+		producerConfig.remove("transactional.id");
```
Comment (pnowojski) [inline flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java]: why do you need this change?

File: flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011ITCase.java [hunk 2]
```diff
@@ -566,6 +568,76 @@ public void testRunOutOfProducersInThePool() throws Exception {
...
+		assertExactlyOnceForTopic(createProperties(), topic, 0, Arrays.asList(42, 43, 44), 30_000L);
+		deleteTestTopic(topic);
+	}
+
+	@Test
+	public void testMigrateFromAtExactlyOnceToAtLeastOnce() throws Exception {
+		String topic = "testMigrateFromExactlyOnceToAtLeastOnce";
+
+		OperatorSubtaskState producerSnapshot;
+		try (OneInputStreamOperatorTestHarness<Integer, Object> testHarness = createTestHarness(topic, EXACTLY_ONCE)) {
```
Comment (pnowojski) [inline flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011ITCase.java]: please deduplicate this code with `testMigrateFromAtLeastOnceToExactlyOnce`

```
private void testMigrateSemantic(Semantic from, Semantic to, String topicName) {
  // both try blocks
}
```

File: flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011ITCase.java [hunk 3]
```diff
@@ -566,6 +568,76 @@ public void testRunOutOfProducersInThePool() throws Exception {
...
+		}
+
+		try (OneInputStreamOperatorTestHarness<Integer, Object> testHarness = createTestHarness(topic, EXACTLY_ONCE)) {
+			testHarness.setup();
+			// restore from snapshot, all records until here should be persisted
+			testHarness.initializeState(producerSnapshot);
+			testHarness.open();
+
+			// write and commit more records
+			testHarness.processElement(44, 7);
```
Comment (pnowojski) [inline flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011ITCase.java]: start this from `45`. Now it's not obvious whether the result `44` comes from here or from above.

File: flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011ITCase.java [hunk 4]
```diff
@@ -566,6 +568,76 @@ public void testRunOutOfProducersInThePool() throws Exception {
...
+			testHarness.snapshot(2, 8);
```
Comment (pnowojski) [inline flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011ITCase.java]: also extend the second try block a little bit by completing at least one checkpoint to make sure that committing also works.

File: flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011ITCase.java [hunk 5]
```diff
@@ -566,6 +568,76 @@ public void testRunOutOfProducersInThePool() throws Exception {
...
+			// write and commit more records, after potentially lingering transactions
+			testHarness.processElement(44, 7);
+			testHarness.snapshot(2, 8);
+			testHarness.processElement(45, 9);
+		}
+
+		//now we should have:
+		// - records 42 and 43 in committed transactions
+		// - aborted transactions with records 44 and 45
+		assertExactlyOnceForTopic(createProperties(), topic, 0, Arrays.asList(42, 43, 44, 45), 30_000L);
```
Comment (pnowojski) [inline flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011ITCase.java]: Could you provide overloaded version of `assertExactlyOnceForTopic` with default value for `long timeoutMillis = 30_000L` as separate commit? I know that it was like that before, but I have only now realised how duplicated (mostly by me) the magic constant `30_000L` is everywhere...

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducerITCase.java [hunk 6]
```diff
@@ -163,7 +163,7 @@ public void testRestoreToCheckpointAfterExceedingProducersPool() throws Exceptio
 	}
 
 	@Test
-	public void testFlinkKafkaProducer10FailBeforeNotify() throws Exception {
+	public void testFlinkKafkaProducerFailBeforeNotify() throws Exception {
```
Comment (pnowojski) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducerITCase.java]: Those changes should be in separate commit.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java [hunk 7]
```diff
@@ -320,9 +320,9 @@ public void initializeState(FunctionInitializationContext context) throws Except
 
 		state = context.getOperatorStateStore().getListState(stateDescriptor);
 
+		boolean recoveredUserContext = false;
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: is this relevant to the bug fix or is this a clean up refactor?

File: flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java (hunk 1 above)
Comment (StefanRRichter) [reply flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java]: Because otherwise the the restore of state from the `EXACTLY_ONCE` case will insert this key in the producer config and we will end up having a producer that is configured for transactions, which we don't want.

File: flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011ITCase.java (hunk 2 above)
Comment (StefanRRichter) [reply flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011ITCase.java]: I think that should work, but the expected result also needs to be provided because it is different.

File: flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011ITCase.java (hunk 3 above)
Comment (StefanRRichter) [reply flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011ITCase.java]: can do

File: flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011ITCase.java (hunk 5 above)
Comment (StefanRRichter) [reply flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011ITCase.java]: A couple of other things in the test might be somewhat duplicated as well, I can just do a general pass and dedup stuff.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java (hunk 7 above)
Comment (StefanRRichter) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: This is relevant, there is a semantical difference. Small, but important.

**Comment by StefanRRichter** [comment]
Addressed the comments. Please have another look.

**Comment by StefanRRichter** [comment]
Thanks for the review @pnowojski. I will keep the hotfix commit separated. Merging.
